src/lib.rs

```diff
@@ -675,13 +930,191 @@ impl Contract {
+                let proposal: Proposal = self
+                    .proposals
+                    .get(proposal_id.into())
+                    .unwrap_or_else(|| panic!("Proposal id {} not found", proposal_id))
+                    .into();
+                if proposal.snapshot.body.latest_version().timeline.is_approved() {
+                    has_approved_proposal = true;
+                    break;
+                }
+            }
+            require!(has_approved_proposal, "Cannot change RFP status to Proposal Selected without an approved proposal linked to this RFP");
+        }
+
+        let old_snapshot = rfp.snapshot.clone();
+        let old_labels_set = old_snapshot.labels.clone();
+        let new_labels = labels;
+        let new_snapshot = RFPSnapshot {
+            editor_id: env::predecessor_account_id(),
+            timestamp: env::block_timestamp(),
+            labels: new_labels.clone(),
+            body: body,
+            linked_proposals: old_snapshot.linked_proposals.clone(),
+        };
+        rfp.snapshot = new_snapshot;
+        rfp.snapshot_history.push(old_snapshot);
+        self.rfps.replace(id.try_into().unwrap(), &rfp.clone().into());
+
+        // Update labels index.
+        let new_labels_set = new_labels;
+
+        let mut edit_proposal_promise: Option<Promise> = None;
+
+        if !old_labels_set.eq(&new_labels_set.clone()) {
+            for proposal_id in self.get_rfp_linked_proposals(id) {
+                edit_proposal_promise = Some(self.update_proposal_labels(proposal_id, new_labels_set.clone()));
+            }
+        }
+
+        let labels_to_remove = &old_labels_set - &new_labels_set;
+        let labels_to_add: HashSet<String> = &new_labels_set - &old_labels_set;
+        require!(
+            self.is_allowed_to_use_labels(
+                Some(editor_id.clone()),
+                labels_to_remove.iter().cloned().collect()
+            ),
+            "Not allowed to remove these labels"
+        );
+        require!(
+            self.is_allowed_to_use_labels(
+                Some(editor_id.clone()),
+                labels_to_add.iter().cloned().collect()
+            ),
+            "Not allowed to add these labels"
+        );
+
+        for label_to_remove in labels_to_remove {
+            let mut rfps = self.label_to_rfps.get(&label_to_remove).unwrap();
+            rfps.remove(&id);
+            self.label_to_rfps.insert(&label_to_remove, &rfps);
+        }
+
+        for label_to_add in labels_to_add {
+            let mut rfps = self.label_to_rfps.get(&label_to_add).unwrap_or_default();
+            rfps.insert(id);
+            self.label_to_rfps.insert(&label_to_add, &rfps);
+        }
+
+        let notify_promise = notify::notify_rfp_subscribers(&rfp, self.get_moderators());
+
+        if let Some(edit_proposal_promise) = edit_proposal_promise {
+            edit_proposal_promise.then(notify_promise)
+        } else {
+            notify_promise
+        }
```

Comment:
I don't think we need to fail the transaction if `notify_edit_proposal` or `notify_rfp_subscribers` fail, so this function can just return `rfp_id` and let the notification promises to resolve without affecting the transaction status.

```suggestion
```